Show all elements in tree view

amolenaar opened the pull request:
### PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->
- [ ] Bugfix
- [X] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Documentation content changes

### What is the current behavior?

Currently only elements with a `NamedElement.namespace` relation are shown in the tree view.

Issue Number: #386 

### What is the new behavior?

The tree view will use the `Element.owner` relation to find all ownership relations. This means (almost) all elements are now visible. A whole lot of elements do not have their owner relation set, however. As a result they end up at root node level in the tree view. This PR does *not* address this issue.

Relationship related elements are grouped under a "<Relationships>" node.

### Does this PR introduce a breaking change?
- [X] Yes
- [ ] No

### Other information

* SysML dependencies like Trace and Refine now inherit from Dependency (they used to inherit from UML.Trace and UML.Refine, which are stereotypes defined in the UML standard profile).
* `Element.ownedComment` has been renamed to `Element.comment`, in line with UML 2.5.1 spec.
* Fixed issue with complex derived union relationships.

![image](https://user-images.githubusercontent.com/96249/92334769-42364b00-f091-11ea-894f-e27f71961de6.png)

danyeaw commented:
I was playing around with the branch and created a generalization relationship between two Classes:
```
Traceback (most recent call last):
  File "/home/dan/Projects/gaphor/gaphor/ui/namespace.py", line 330, in sort_func
    b = (format(vb) or "").lower()
  File "/home/dan/.pyenv/versions/3.8.5/lib/python3.8/functools.py", line 875, in wrapper
    return dispatch(args[0].__class__)(*args, **kw)
  File "/home/dan/Projects/gaphor/gaphor/UML/umlfmt.py", line 206, in format_generalization
    return f"general: {el.general.name}"
AttributeError: 'NoneType' object has no attribute 'name'
```

danyeaw reviewed (approved):
Really nice updates, this greatly improves the tree view to show relationships and makes it much more apparent that there is a data model!

danyeaw commented:
I can reproduce this AttributeError.

1. Create two classes
2. Create a composite association between them
3. Expand Relationships in the tree view
4. Create a generalization between the two Classes in step 1.

amolenaar commented:
Nice catch. I will fix this one. 

Currently it's still possible to drag pretty much all elements on the canvas. I think for now I will put something in place to prevent this (e.g. action flows) to avoid inconsistency between model and diagrams.

amolenaar commented:
Cleanup is not done properly.

![image](https://user-images.githubusercontent.com/96249/92397962-f5ea1a00-f127-11ea-8c5a-2fe9001f988e.png)

amolenaar commented on gaphor/core/format.py:
```diff
@@ -0,0 +1,26 @@
+    """
+    Format an element.
+    """
+    raise TypeError("Format routine for type %s not implemented yet" % type(el))
```
Indeed that would be nicer.

amolenaar commented:
> Cleanup is not done properly.
> 
> Those "general" entries in the tree view are not backed by a model element.

Fixed.

I also introduced a NamespaceModel class, so the model updating code from the Namespace UI component could be moved to a slightly smaller class. I think this enhances the readability already, although the classes are still part of the same module.

danyeaw commented:
Sorry to keep on breaking stuff :smile: :bug: 
```
Traceback (most recent call last):
  File "/home/dan/Projects/gaphor/gaphor/ui/namespace.py", line 568, in _on_view_event
    menu = Gtk.Menu.new_from_model(self.namespace_popup_model())
  File "/home/dan/Projects/gaphor/gaphor/ui/namespace.py", line 544, in namespace_popup_model
    gettext('Show in "{diagram}"').format(diagram=diagram.name),
AttributeError: 'NoneType' object has no attribute 'name'
```
To reproduce:
1. Create two classes
2. Create a composite association between them
3. Right click on the head or tail end in the tree view

danyeaw commented on gaphor/ui/namespace.py:
```diff
@@ -248,153 +271,88 @@ def on_drag_data_received(self, context, x, y, selection, info, time):
         """Get the Gtk.TreeIter for an element in the Namespace.
 
         Args:
             element: The element contained in the in the Namespace.
```
Duplicative "in the"

danyeaw commented on gaphor/UML/classes/klass.py:
```diff
@@ -177,7 +178,7 @@ def attributes_compartment(subject):
     # We need to fix the attribute value, since the for loop changes it.
     def lazy_format(attribute):
         # str(), so we never ever get an error on a property part of an association
```
Is this str() comment still valid?

danyeaw reviewed (approved):
This looks really great, just a few nitpick comments.
Should we release version 2.1 once this is merged?

danyeaw commented:
I realized that the latest attribute error exists in version 2.0.1, so I don't think we need to fix that as part of this PR. I created #438 to track.

amolenaar commented:
> This looks really great, just a few nitpick comments.
> Should we release version 2.1 once this is merged?

Well... a lot of elements are not assigned a proper owner, so they will all appear in the root. I want to make sure the elements end up in the right package or on their respective owner element first.

amolenaar commented on gaphor/core/format.py:
```diff
@@ -0,0 +1,21 @@
+
+
+@singledispatch
+def format(el: Element) -> str:
```
It does overload the builtin `format` function only when imported. It used to be called `format`, so I left it. A more descriptive name would be better, I guess...

amolenaar commented:
@danyeaw I fixed the comments you made. Except for the one on the `format` function. We may want to decide on a new name, if any.

danyeaw commented:
@amolenaar Thanks for accepting my feedback! I think keeping it as format is fine for now, we'll just have to be careful that it doesn't ever clash with the builtin name. Amazing job again on this, I'm really excited to do some modeling and see all the relationships in the tree view.
